analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs

```diff
@@ -136,65 +132,114 @@ private static List<Location> GetDuplicatedCastLocations(SyntaxNodeAnalysisConte
+            else
             {
-                for (var i = 0; i < recursivePositionalPattern.PositionalPatternClause.Subpatterns.Count; i++)
+                AddPatternType(pattern, targetTypes);
+            }
+            return targetTypes;
+
+            static void AddPatternType(SyntaxNode pattern, ISet<TypeSyntax> targetTypes)
+            {
+                if (GetType(pattern) is { } patternType)
                 {
-                    var pattern = recursivePositionalPattern.PositionalPatternClause.Subpatterns[i].Pattern;
-                    if (ReportDeclarationPatternAndReturnType(analysisContext, isPatternLocation, pattern, parentStatement) is { } patternType
-                        && mainVariableExpression.IsKind(SyntaxKindEx.TupleExpression)
-                        && (TupleExpressionSyntaxWrapper)mainVariableExpression is var tupleExpression
-                        && tupleExpression.Arguments.Count == recursivePositionalPattern.PositionalPatternClause.Subpatterns.Count)
-                    {
-                        ReportPatternAtMainVariable(analysisContext, tupleExpression.Arguments[i].Expression, mainVariableLocation, parentStatement, patternType, message);
-                    }
+                    targetTypes.Add(patternType);
                 }
             }
         }
 
-        private static TypeSyntax ReportDeclarationPatternAndReturnType(SyntaxNodeAnalysisContext analysisContext, Location isPatternLocation, SyntaxNode pattern, SyntaxNode parentIfStatement)
+        private static TypeSyntax GetType(SyntaxNode pattern)
         {
-            if (pattern.IsKind(SyntaxKindEx.DeclarationPattern)
-                && (DeclarationPatternSyntaxWrapper)pattern is var declarationPattern
-                && declarationPattern.Designation.SyntaxNode.IsKind(SyntaxKindEx.SingleVariableDesignation)
-                && (SingleVariableDesignationSyntaxWrapper)declarationPattern.Designation.SyntaxNode is var singleVariableDesignation)
+            if (ConstantPatternSyntaxWrapper.IsInstance(pattern))
+            {
+                return ((ConstantPatternSyntaxWrapper)pattern).Expression as TypeSyntax;
+            }
+            else if (DeclarationPatternSyntaxWrapper.IsInstance(pattern))
             {
-                ReportPatternAtCastLocation(analysisContext, singleVariableDesignation.SyntaxNode, isPatternLocation, parentIfStatement, declarationPattern.Type, RemoveRedundantCaseMessage);
-                return declarationPattern.Type;
+                return ((DeclarationPatternSyntaxWrapper)pattern).Type;
+            }
+            else if (RecursivePatternSyntaxWrapper.IsInstance(pattern))
+            {
+                return ((RecursivePatternSyntaxWrapper)pattern).Type;
             }
             return null;
         }
 
+        private static void MapObjectToPattern(ExpressionSyntax expression, SyntaxNode pattern, IDictionary<ExpressionSyntax, SyntaxNode> objectToPatternMap)
```

Comment:
Now that S3981 PR is merged, please check that the two methods are still the same and move it two helpers.